### ISSUE TITLE
Add rule to ban compound selectors whose complex selectors contain vendored selectors from different vendors

### ISSUE THREAD

**Issue by Goodwine** [stylelint/stylelint#8588]
### What is the problem you're trying to solve?

That was a mouthful!

The issue I'm trying to solve is documented at:
- https://css-tricks.com/one-invalid-pseudo-selector-equals-an-entire-ignored-selector/

The problem appears in cases such as this one:
```css
.input::-webkit-input-placeholder,
.input::-moz-placeholder,
.input::placeholder {
  color: red;
}
```

In this case, Chrome will not recognize `::-moz`, and because of how browsers work, this invalidates the selector `::-moz-placeholder` which in turns invalidate the whole style rule. This happens silently and in most cases users won't realize this happens.

As a side-note, Autoprefixer may sometimes hide this issue if it happens to create a duplicate rule. In the given example above, it actually hides the issue for "Chrome >=56" but not for other target browsers 😛 

### What solution would you like to see?

I think that stylelint could introduce a check that prohibits the mixing of vendored pseudo selectors. Note that this would be a different rule than [selector-no-vendor-prefix](https://stylelint.io/user-guide/rules/selector-no-vendor-prefix/)

---
# Good Examples
All separated
```scss
.input::-webkit-input-placeholder{
  color: red;
}
.input::-moz-placeholder{
  color: red;
}
.input::placeholder {
  color: red;
}
```

Only vendored selectors are combined
```scss
.input::-webkit-input-placeholder,
::-webkit-scrollbar {
  color: red;
}
.input::-moz-placeholder,
::-moz-scrollbar {
  color: red;
}
.input::placeholder {
  color: red;
}
```

---

# Bad Examples

Combines vendored selectors with non-vendored selectors

```scss
.input::-webkit-input-placeholder,
.input::placeholder {
  color: red;
}
```

Combines vendored selectors from different vendors

```scss
.input::-webkit-input-placeholder,
.input::-moz-input-placeholder {
  color: red;
}
```

---

# Fix

Fixing the issue should be straightforward by duplicating the rule such that the "compound selector" is split and grouped according to whether the "complex selectors" have the same vendor or not.

**Comment by MorevM** [comment]
Good day!
I ran into the same problem today, so here I am :)

UPD: I just found that `stylelint-plugin-defensive-css` already provides a [`no-mixed-vendor-prefixes`](https://github.com/yuschick/stylelint-plugin-defensive-css/tree/main#no-mixed-vendor-prefixes) rule.
Its implementation is simpler than the one I'm proposing, but it resolves most of the issues.

I'll move my original post under the `details` section instead of deleting, since I still think that implementation would be valuable in Stylelint core. But given the context described above, if there's an alternative, it's unlikely to happen :)
Maybe the context I've gathered might help in making a decision - the issue is still open. 

<details>
<summary>Show original post</summary>
My scenario was:

```css
:where(input[type='range'])::-webkit-slider-thumb,
:where(input[type='range'])::-moz-range-thumb {
 // ...
}
```

In Blink/WebKit, the unknown `::-moz-range-thumb` invalidates the whole selector list. `:where()` does not help because its forgiving behavior applies only to its own arguments, and pseudo-elements cannot be placed inside it.

I'd like to revisit this proposal, as I was surprised that Stylelint did not catch this error for me.

---

I do not think the proposed rule would conflict with [`selector-no-vendor-prefix`](https://stylelint.io/user-guide/rules/selector-no-vendor-prefix/).
The rule only reports selectors recognized by [`isAutoprefixable.selector()`](https://github.com/stylelint/stylelint/blob/main/lib/rules/selector-no-vendor-prefix/index.mjs#L55-L57). The helper checks against a [finite list of selectors](https://github.com/stylelint/stylelint/blob/main/lib/utils/isAutoprefixable.mjs#L36-L52) that can be safely replaced and handled by tools like Autoprefixer.

Neither `::-webkit-slider-thumb` nor `::-moz-range-thumb` is included, and neither can be fixed simply by removing its prefix.

---

The ongoing [CSS Form Control Styling Level 1](https://drafts.csswg.org/css-forms-1/) also makes the problem more relevant. It introduces standard parts such as:

* [`::slider-thumb`, `::slider-track`, and `::slider-fill`](https://drafts.csswg.org/css-forms-1/#slider-pseudos);
* [`::step-control`, `::step-up`, and `::step-down`](https://drafts.csswg.org/css-forms-1/#number-pseudos);
* [`::field-component` and `::field-separator`](https://drafts.csswg.org/css-forms-1/#date-time-pseudos).

These parts are associated with [`appearance: base`](https://drafts.csswg.org/css-forms-1/#appearance). The specification explicitly states that existing vendor pseudo-elements should be implemented as aliases [only when possible](https://drafts.csswg.org/css-forms-1/#:~:text=When%20not%20possible%2C%20the%20user%20agent%20must%20reserve%20the%20standard%20pseudo%2Delements). 

This means there is not necessarily a one-to-one mapping to unprefixed version, so authors may **still need to write vendor-specific selectors manually**.

---

I can see this as a core rule.

The check itself could remain narrow: mixing different vendor prefixes should be reported, as should mixing a non-`-webkit-` vendor prefix with an unprefixed pseudo-element. Unknown non-functional [`::-webkit-*` pseudo-elements are valid](https://drafts.csswg.org/selectors/#compat), so they do not cause the same problem.


```css
/* Report */ 
input::-webkit-slider-thumb, 
input::-moz-range-thumb {} 

/* Report */ 
input::-moz-range-thumb, 
input::slider-thumb {} 

/* Do not report */ 
input::-webkit-slider-thumb, 
input::-webkit-slider-runnable-track {}
```

---

Does the above change your opinion in any way?
The problem seems general and important enough for Stylelint core, and I would be happy to implement it.
</details>

**Comment by ybiquitous** [comment]
@MorevM Thanks for sharing the plugin. From the discussion, I'm leaning toward adding a new built-in rule so that authors can use a *forgiving selector list* instead.

See also the [CSS Selector spec](https://drafts.csswg.org/selectors/#forgiving-selector) that mentions the benefits of `<forgiving-selector-list>`.

A new rule might be as below (there might be a better name):

```json
{
  "selector-no-unforgiving-list": true
}
```

```css
/* NG ❌ */
.input::-webkit-input-placeholder,
.input::-moz-placeholder,
.input::placeholder {
  color: red;
}

/* OK ✅ */
:is(
  .input::-webkit-input-placeholder,
  .input::-moz-placeholder,
  .input::placeholder
) {
  color: red;
}
```

To prevent too many problems with this rule, it would initially be scoped only to mixed-vendor pseudo-selectors.

> [!NOTE]
> There was a similar discussion in #6524 about the forgiving selector list.

This is just my rough idea and has to be refined. Please give any feedback.

**Comment by MorevM** [comment]
@ybiquitous Good day!

That might be a good idea, but unfortunately, `:is`, `:where`, and `:has` [do not accept pseudo-elements as arguments](https://drafts.csswg.org/selectors/#matches:~:text=Pseudo%2Delements%20cannot%20be%20represented%20by%20the%20matches%2Dany%20pseudo%2Dclass%3B%20they%20are%20not%20valid%20within%20%3Ais().).

```scss
// This is syntactically correct, 
// but never selects anything.

:is(::placeholder) {}
```

This, by the way, is a good extension for `selector-no-invalid`.
Currently, the rule does not report an error. I can do this if you don't mind.

**Comment by ybiquitous** [comment]
@MorevM Oh, I missed the important notice in the spec I mentioned. Great catch!

The spec says:

> so long as it doesn’t contain any pseudo-elements (which aren’t valid in `:is()` or `:where()`)

So, I'm turning back to my original idea. That is, this proposed idea should be implemented by a plugin, not by a built-in rule. I understand that some vendor prefixes are still necessary in such a situation, but they're contrary to our current direction.

**Comment by ybiquitous** [comment]
@MorevM By the way, your proposal below sounds great to me. Could you open another issue, please? 👍🏼 

> This, by the way, is a good extension for `selector-no-invalid`.
> Currently, the rule does not report an error. I can do this if you don't mind.

The title would be like so:

```markdown
Fix `selector-no-invalid` false negatives for pseudo-elements in `:is()` and `:where()`
```

**Comment by jeddy3** [comment]
> So, I'm turning back to my original idea. That is, this proposed idea should be implemented by a plugin, not by a built-in rule. I understand that some vendor prefixes are still necessary in such a situation, but they're contrary to our current direction.

Closing as best handled in community plugins/
